File: apps/web-console/src/app/runtime-config.ts
/**
 * Runtime configuration fetched from /ui/runtime-config.json at startup.
 * This allows a single immutable image to be deployed across environments —
 * the container/Helm chart injects the JSON; we never hardcode these values
 * at build time via import.meta.env.
 */
export interface RuntimeConfig {
  apiBaseUrl: string;
  keycloakUrl: string;
  keycloakRealm: string;
  keycloakClientId: string;
  tenantBanner?: string;
}

let _config: RuntimeConfig | null = null;

export async function loadRuntimeConfig(): Promise<RuntimeConfig> {
  const response = await fetch('/ui/runtime-config.json');
  if (!response.ok) {
    throw new Error(`Failed to load runtime config: ${response.status}`);
  }
  _config = await response.json() as RuntimeConfig;
  return _config;
}

export function getRuntimeConfig(): RuntimeConfig {
  if (!_config) {
    throw new Error('Runtime config not loaded — call loadRuntimeConfig() first');
  }
  return _config;
}

File: apps/control-api/app/api/routers/runtime_config.py
from fastapi import APIRouter

from app.api.schemas.runtime_config import UiRuntimeConfig
from app.settings import settings

router = APIRouter(include_in_schema=False)


@router.get("/ui/runtime-config.json", response_model=UiRuntimeConfig)
async def get_runtime_config() -> UiRuntimeConfig:
    return UiRuntimeConfig(
        apiBaseUrl=settings.public_base_url,
        keycloakUrl=settings.keycloak_browser_url,
        keycloakRealm=settings.keycloak_browser_realm,
        keycloakClientId=settings.keycloak_browser_client_id,
        tenantBanner=settings.tenant_banner,
    )

